src/libstd/sys/unix/fs.rs

```diff
@@ -463,18 +468,24 @@ pub fn rmdir(p: &Path) -> io::Result<()> {
 pub fn readlink(p: &Path) -> io::Result<PathBuf> {
     let c_path = try!(cstr(p));
     let p = c_path.as_ptr();
-    let mut len = unsafe { libc::pathconf(p as *mut _, libc::_PC_NAME_MAX) };
-    if len < 0 {
-        len = 1024; // FIXME: read PATH_MAX from C ffi?
-    }
-    let mut buf: Vec<u8> = Vec::with_capacity(len as usize);
-    unsafe {
-        let n = try!(cvt({
-            libc::readlink(p, buf.as_ptr() as *mut c_char, len as size_t)
-        }));
-        buf.set_len(n as usize);
+
+    let mut buf = Vec::with_capacity(256);
+
+    loop {
+        let buf_read = try!(cvt(unsafe { libc::readlink(
+                                                    p,
+                                                    buf.as_mut_ptr() as *mut libc::c_char,
+                                                    buf.capacity() as libc::size_t) })) as usize;
+        if buf_read != buf.capacity() {
+            unsafe { buf.set_len(buf_read); }
+            buf.shrink_to_fit();
+
+            return Ok(PathBuf::from(OsString::from_vec(buf)));
+        }
+
+        let new_cap = buf.capacity() * 2;
+        buf.reserve(new_cap);
```

Comment:
Uhm, sorry for asking this again, but there's still a remaining concern. As the length is not actually set to the capacity at each iteration currently, calling `buf.reserve(1)` will be a no-op in my code.

Do you think the length should be updated every time before calling `reserve`? Something like:

```rust
loop {
    let buf_read = try!(readlink(...));

    unsafe { buf.set_len(buf_read); } // This will be added

    if buf_read != buf.capacity() {
        unsafe { buf.set_len(buf_read); } // This should probably be removed
        buf.shrink_to_fit();

        return Ok(...);
    }

    buf.reserve(1);
}
```

That feels a bit odd to me because unlike `read_to_end_uninitialized`, we abandon the buffer immediately if `readlink` fails. But if you're OK with the code above, I'll follow that.

Plus, do you think [getcwd](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/os.rs#L101-L114) might also have to be modified to use `reserve(1)`? This may have to be another PR, but if you like it, I can add another commit to address this.

Sorry for the disturbance, this shall be my last review request!